myia/abstract/utils.py

```diff
@@ -745,122 +753,128 @@ def amerge(x1: dtype.TypeMeta, x2, loop, forced, bp):
     return AbstractTuple(*merged)
 
 
 @overload  # noqa: F811
-def amerge(x1: AbstractArray, x2, loop, forced, bp):
+def amerge(self, x1: AbstractArray, x2, forced, bp):
     args1 = (x1.element, x1.values)
     args2 = (x2.element, x2.values)
-    merged = amerge(args1, args2, loop, forced, bp)
+    merged = self(args1, args2, forced, bp)
     if forced or merged is args1:
         return x1
     return AbstractArray(*merged)
 
 
 @overload  # noqa: F811
-def amerge(x1: AbstractList, x2, loop, forced, bp):
+def amerge(self, x1: AbstractList, x2, forced, bp):
     args1 = (x1.element, x1.values)
     args2 = (x2.element, x2.values)
-    merged = amerge(args1, args2, loop, forced, bp)
+    merged = self(args1, args2, forced, bp)
     if forced or merged is args1:
         return x1
     return AbstractList(*merged)
 
 
 @overload  # noqa: F811
-def amerge(x1: AbstractClassBase, x2, loop, forced, bp):
+def amerge(self, x1: AbstractClassBase, x2, forced, bp):
     args1 = (x1.tag, x1.attributes, x1.methods, x1.values)
     args2 = (x2.tag, x2.attributes, x2.methods, x2.values)
-    merged = amerge(args1, args2, loop, forced, bp)
-    if forced or merged is args1:
+    merged = self(args1, args2, forced, bp)
+    if forced or merged is args1:  # pragma: no cover
+        # This used to be covered before the `x1 is x2` check in amerge
         return x1
     return type(x1)(*merged)
 
 
 @overload  # noqa: F811
-def amerge(x1: AbstractJTagged, x2, loop, forced, bp):
+def amerge(self, x1: AbstractJTagged, x2, forced, bp):
     args1 = x1.element
     args2 = x2.element
-    merged = amerge(args1, args2, loop, forced, bp)
-    if forced or merged is args1:
+    merged = self(args1, args2, forced, bp)
+    if forced or merged is args1:  # pragma: no cover
+        # This used to be covered before the `x1 is x2` check in amerge
         return x1
     return AbstractJTagged(merged)
 
 
 @overload  # noqa: F811
-def amerge(x1: (AbstractUnion, AbstractTaggedUnion), x2, loop, forced, bp):
+def amerge(self, x1: (AbstractUnion, AbstractTaggedUnion),
+           x2, forced, bp):
     args1 = x1.options
     args2 = x2.options
-    merged = amerge(args1, args2, loop, forced, bp)
-    if forced or merged is args1:
+    merged = self(args1, args2, forced, bp)
+    if forced or merged is args1:  # pragma: no cover
+        # Covered in the next PR
         return x1
     return type(x1)(merged)
 
 
 @overload  # noqa: F811
-def amerge(x1: (int, float, bool), x2, loop, forced, bp):
+def amerge(self, x1: (int, float, bool), x2, forced, bp):
     if forced and x1 != x2:
         raise TypeMismatchError(x1, x2)
     return x1 if x1 == x2 else ANYTHING
 
 
 @overload  # noqa: F811
-def amerge(x1: object, x2, loop, forced, bp):
+def amerge(self, x1: object, x2, forced, bp):
     if x1 != x2:
         raise TypeMismatchError(x1, x2)
-    return x1
+    else:  # pragma: no cover
+        # This used to be covered before the `x1 is x2` check in amerge
+        return x1
```

Comment:
This case should be relatively easy to cover with something like `amerge([1], [1])`